File: JavaScript/611-Valid Triangle Number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var triangleNumber = function(nums) {
    nums.sort((a,b) => a-b);
    let n = nums.length;
    let res = 0;
    
    for (let k=n-1; k>=2; k--){
        let i=0, j=k-1;
        while (i<j){
            if (nums[i] + nums[j] > nums[k]){
                res += j-i;
                j--;
            }else{
                i++;
            }
        }
    }
    return res;
};

File: C++/611-Valid Triangle Number.cpp
class Solution {
public:
    int triangleNumber(vector<int>& nums) {
        sort(begin(nums), end(nums));
        int N = nums.size(), res = 0;

        for (int i=N-1; i>=2; i--) {
            int left = 0, right = i-1;
            while (left < right) {
                if (nums[left] + nums[right] > nums[i]) {
                    res += right-left;
                    right--;
                } else {
                    left++;
                }
            }
        }

        return res;
    }
};